index.js

```diff
@@ -215,16 +66,100 @@ var publicAgent = {
  * @example
  * trace.start();
  */
-function start(config) {
-  publicAgent.start(config);
-  return publicAgent;
+function start(projectConfig) {
+  var config = initConfig(projectConfig);
+
+  if (traceApi.isActive() && !config.forceNewAgent_) { // already started.
+    throw new Error('Cannot call start on an already started agent.');
+  }
+
+  if (!config.enabled) {
+    return traceApi;
+  }
+
+  var logLevel = config.logLevel;
+  if (logLevel < 0) {
+    logLevel = 0;
+  } else if (logLevel >= common.logger.LEVELS.length) {
+    logLevel = common.logger.LEVELS.length - 1;
+  }
+  var logger = common.logger({
+    level: common.logger.LEVELS[logLevel],
+    tag: '@google/cloud-trace'
+  });
+
+  if (!semver.satisfies(process.versions.node, '>=0.12')) {
+    logger.error('Tracing is only supported on Node versions >=0.12');
+    return traceApi;
+  }
+
+  if (config.projectId) {
+    logger.info('Locally provided ProjectId: ' + config.projectId);
+  }
+
+  if (onUncaughtExceptionValues.indexOf(config.onUncaughtException) === -1) {
+    logger.error('The value of onUncaughtException should be one of ',
+      onUncaughtExceptionValues);
+    throw new Error('Invalid value for onUncaughtException configuration.');
+  }
+
+  var headers = {};
+  headers[constants.TRACE_AGENT_REQUEST_HEADER] = 1;
+
+  if (modulesLoadedBeforeTrace.length > 0) {
+    logger.warn('Tracing might not work as the following modules ' +
+      'were loaded before the trace agent was initialized: ' +
+      JSON.stringify(modulesLoadedBeforeTrace));
+  }
+
+  if (typeof config.projectId === 'undefined') {
+    // Queue the work to acquire the projectId (potentially from the
+    // network.)
+    gcpMetadata.project({
+      property: 'project-id',
+      headers: headers
+    }, function(err, response, projectId) {
+      if (response && response.statusCode !== 200) {
+        if (response.statusCode === 503) {
+          err = new Error('Metadata service responded with a 503 status ' +
+            'code. This may be due to a temporary server error; please try ' +
+            'again later.');
+        } else {
+          err = new Error('Metadata service responded with the following ' +
+            'status code: ' + response.statusCode);
+        }
+      }
+      if (err) {
+        logger.error('Unable to acquire the project number from metadata ' +
+          'service. Please provide a valid project number as an env. ' +
+          'variable, or through config.projectId passed to start(). ' + err);
+        if (traceApi.isActive()) {
+          agent.stop();
+          traceApi.disable_();
+          pluginLoader.deactivate();
+        }
+        return;
+      }
+      config.projectId = projectId;
+    });
+  } else if (typeof config.projectId !== 'string') {
+    logger.error('config.projectId, if provided, must be a string. ' +
+      'Disabling trace agent.');
+    return traceApi;
+  }
+
+  agent = require('./src/trace-agent.js').get(config, logger);
```

Comment:
Any reason why this require is not done at the top?

Secondly, we have three variables here that have a strong dependence on each other. `agent`, `traceApi`, and `pluginLoader`. 

* First of all can we move `pluginLoader` into the agent? it should be a concept owned by the agent. Since there is a strong dependence on `pluginLoader.activate` and `deactivate`, when `agent.start` and `stop` are called respectively, it would better if `agent` was *responsible* for calling these? I see that you moved this from the agent to here. Is there a reason for doing this?
* Similarly, can we make the `agent` responsible for calling `traceApi.enable` and `disable` from `start` and `stop` respectively?
* Nit: we have three names for similar operations: `{start, enable, activate}` and `{stop, disable, deactivate}`. Do we really need all three?